[PORT, NOVHF] CHECKDATASIG and CHECKDATASIGVERIFY

awemany opened the pull request:
This is the implementation of `CHECKDATASIG/-VERIFY` (CDS/-V) from Bitcoin ABC, ported onto Bitcoin Unlimited.

A few notes for reviewers:

- ~~includes #1389~~ (not anymore as this is back to WIP)
- ABC's code has been taken as much as possible to keep the chances of introducing a chain split bug low
- The old DSV code has been removed and then replaced iteratively with what is on ABC
- I understand the reason to go with a global `bool` flag instead of passing another flag parameter, but I am not at all sure about the sequencing and initialization code with regards to MTP as well as any conflicting accesses by multiple threads (block validation, mempool admission) that modify the global flag. Reviewers, please ensure that the implemented logic is sound.
- `sighashtype.h` has been ported (from HEAD of ABC) to BU to aid in porting tests - but isn't used anywhere else
- The code relies on signatures that are encoded without the final `sighash` byte. To keep changes to critical and divergent code paths in BU low, I have opted to include an additional `IsValidSignatureEncodingWithoutSigHash()` in `interpreter.cpp`, which is ABC's current implementation. This is a TODO item still, it might make sense to integrate both and base everything off the `IsValidSignatureEncodingWithoutSigHash(..)` function. Overall, this lead to ugly code duplication and I am not at all happy with the current state here.
- a check for compressed pubkeys has been ported from Core as this is used as an obligatory criterion for pubkeys used in CDS/-V, plus tests for this and tests for `CheckSignatureEncoding(..)`
- the exhaustive checking of flag and sighash combinations has been replaced with a randomized check. Note that Amaury opted to replace the sighash testing with just the simplified signature encoding function that does not include the sighash type.
- sigops counting has not been touched in BU, except for just adding the CDS/-V counts when activated
- `script_tests.cpp` which produces the test JSON  file `script_tests.json` has been adapted to have a dummy CHECKDATASIG-enabled flag, to keep the rest of the logic intact. This is somewhat ugly and due to the above change to a global var, though could of course be cleaned up separately.

jl2012 committed:
Import IsCompressedKey from Bitcoin Core

Committer's note: This is a port of the IsCompressedPubKey(..)
that was introduced by Johnson Lau into Bitcoin Core with commit
4c0c25a604cec39675187507df423cff5d707ab5.

The author is set in the commit to keep the attribution.

deadalnix committed:
Decouple non compressed pubkey from segwit

Summary: As per title. This is a useful check in itself and there is no reason to tie to segwit that way.

Test Plan:
  make check

Reviewers: freetrader, #bitcoin_abc

Reviewed By: freetrader, #bitcoin_abc

Differential Revision: https://reviews.bitcoinabc.org/D56

deadalnix committed:
Add CHECKDATASIG opcode.

Summary: The opcode is constantly disabled.

Test Plan: Add test cases to ensure the opcode is disabled.

Reviewers: #bitcoin_abc, schancel

Reviewed By: #bitcoin_abc, schancel

Subscribers: schancel, teamcity

Differential Revision: https://reviews.bitcoinabc.org/D1563

Committer's notice (@awemany):

This has been heavily reworked to fit onto the diverging
Bitcoin Unlimited code base.

Co-authored-by: Awemany <awemany@protonmail.com>

deadalnix committed:
Add a flag to activate OP_CHECKDATASIG

Summary:
As per title.

Depends on D1563

Test Plan: Add test cases in script_test.json and checkdatasig_tests.cpp

Reviewers: #bitcoin_abc, schancel

Reviewed By: #bitcoin_abc, schancel

Subscribers: schancel, teamcity

Differential Revision: https://reviews.bitcoinabc.org/D1564

Committer's notice: Adapted for BU.

Co-authored-by: Awemany <awemany@protonmail.com>

deadalnix committed:
Add test for CheckPubKeyEncoding

Summary:
As per title. More test is better.

Depends on D1565

Test Plan:
  make check

Reviewers: #bitcoin_abc, jasonbcox

Reviewed By: #bitcoin_abc, jasonbcox

Subscribers: jasonbcox, teamcity

Differential Revision: https://reviews.bitcoinabc.org/D1566

Committer's notice:

This has been reworked somewhat. The 'refactors' ({} - brace changes,
size_t -> uint32_t, presumably for ABC's formatting/coding
guidelines) to other files have been dropped from this commit.
Also, header files have been adapted for BU as well as
exporting CheckPubKeyEncoding from interpreter.cpp.

Co-authored-by: Awemany <awemany@protonmail.com>

deadalnix committed:
Add test cases for CheckSignatureEncoding

Summary:
Mostly covering the SigHashType handling so far. More tests required, but it's better than nothing.

Depends on D1566

Test Plan:
  make check

Reviewers: #bitcoin_abc, jasonbcox

Reviewed By: #bitcoin_abc, jasonbcox

Subscribers: teamcity

Differential Revision: https://reviews.bitcoinabc.org/D1567

Committer's notice: The style changes in sigencoding.cpp have been
dropped. Furthermore (FIXME, TODO!), the test for the error flags
SCRIPT_ERR_MUST_USE_FORKID and SCRIPT_ERR_ILLEGAL_FORKID has been
commented out as BU checks the ForkId outside of the
CheckSignatureEncoding(..) function. The whole test case likely
needs to be reworked further to make more sense wrt. the BU codebase.

Co-authored-by: Awemany <awemany@protonmail.com>

deadalnix committed:
Check error code return by CheckPubKeyEncoding in tests

Summary:
This ensure that any change of behavior related to these error code will be detected.

Depends on D1566

Test Plan: make check

Reviewers: #bitcoin_abc, jasonbcox

Reviewed By: #bitcoin_abc, jasonbcox

Subscribers: teamcity

Differential Revision: https://reviews.bitcoinabc.org/D1568

deadalnix committed:
Add test case for CheckSignatureEncoding for proper DER encoding

Summary:
This ensures IsValidSignatureEncoding is properly checked.

The test time is atrocious because IsValidSignatureEncoding assumes the presence of the sighash type which causes a combinotary explosion, and also uncover a SRP violation. I would like to refactor these to detangle the two, but obviously doing so safely require a test so here we are.

Depends on D1570

Test Plan:
  make check

Reviewers: #bitcoin_abc, jasonbcox

Reviewed By: #bitcoin_abc, jasonbcox

Subscribers: jasonbcox, teamcity

Differential Revision: https://reviews.bitcoinabc.org/D1571

deadalnix committed:
Add test for CheckSignatureEncoding to verify low S is checked properly.

Summary:
Increase the unit test coverage for CheckSignatureEncoding.

Depends on D1567

Test Plan:
  make check

Reviewers: #bitcoin_abc, jasonbcox

Reviewed By: #bitcoin_abc, jasonbcox

Subscribers: jasonbcox, teamcity

Differential Revision: https://reviews.bitcoinabc.org/D1570

deadalnix committed:
Add CheckDataSignatureEncoding to check signature with a sighashtype.

Summary:
As per title.

Depends on D1578

Test Plan:
  make check

Reviewers: #bitcoin_abc, jasonbcox

Reviewed By: #bitcoin_abc, jasonbcox

Subscribers: teamcity

Differential Revision: https://reviews.bitcoinabc.org/D1579

Committer's notice: This is basically just taking the test changes from
the corresponding ABC commit.
The actual CheckDataSignatureEncoding has been adapted considerably to BU
due to diverging code bases.

Co-authored-by: Awemany <awemany@protonmail.com>

deadalnix committed:
Update sigopcount_tests.cpp to recent ABC version

Updates sigopcount_tests.cpp to ABC version at
rev.:
b35b4d5a77f5187614f09cc3a6882955a672edcb

Committer's notice: Author of this commit has been
set to the main author of this changeset for attribution.
Some minor tweaks (header files etc.) for BU are included.

Co-authored-by: Awemany <awemany@protonmail.com>

deadalnix committed:
Pass flags down to GetP2SHSigOpCount and GetSigOpCountWithoutP2SH

Summary:
This will ensure we can use flags to tweak the way sigops are counted in the future. That is necessary for the activation of OP_CHECKDATASIG .

Depends on D1596

Test Plan: Added test cases for the logic change.

Reviewers: #bitcoin_abc, jasonbcox

Reviewed By: #bitcoin_abc, jasonbcox

Subscribers: teamcity

Differential Revision: https://reviews.bitcoinabc.org/D1597

Committer's note: Reworked considerable to adapt to the BU codebase.

Co-authored-by: Awemany <awemany@protonmail.com>

deadalnix committed:
Make sure CScript::GetSigOpCount(CScript &) is tested when called on non P2SH scripts.

Summary: This codepath is not tested at the moment.

Test Plan: Added test cases.

Reviewers: #bitcoin_abc, jasonbcox

Reviewed By: #bitcoin_abc, jasonbcox

Subscribers: teamcity

Differential Revision: https://reviews.bitcoinabc.org/D1599

deadalnix committed:
Count OP_CHECKDATASIG and OP_CHECKDATASIGVERIFY as sigops when the proper flag is set.

Summary:
As per title. Once the opcode activates, it needs to count as a sigops.

Depends on D1601

Test Plan: Added test cases.

Reviewers: #bitcoin_abc, schancel, jasonbcox

Reviewed By: #bitcoin_abc, schancel, jasonbcox

Subscribers: jasonbcox, teamcity

Differential Revision: https://reviews.bitcoinabc.org/D1605

Committer's notice: Adapted for BU.

Co-authored-by: Awemany <awemany@protonmail.com>

deadalnix committed:
Implement OP_CHECKDATASIG and OP_CHECKDATASIGVERIFY

Summary:
As per title.

Depends on D1619, D1620 and D1605

Test Plan: Added a ton of test cases.

Reviewers: #bitcoin_abc, jasonbcox

Reviewed By: #bitcoin_abc, jasonbcox

Subscribers: jasonbcox, teamcity

Differential Revision: https://reviews.bitcoinabc.org/D1621

Committer's notice: Reworked to fit onto BU's code base.

Co-authored-by: Awemany <awemany@protonmail.com>

deadalnix committed:
Refactor script_tests.cpp to separate signature generation from PushSig

Summary: It'll allow for addition of test for OP_CHECKDATASIG .

Test Plan:
  make check

Reviewers: #bitcoin_abc, jasonbcox

Reviewed By: #bitcoin_abc, jasonbcox

Subscribers: teamcity

Differential Revision: https://reviews.bitcoinabc.org/D1619

Committer's notice: Adapted to BU.

Co-authored-by: Awemany <awemany@protonmail.com>

deadalnix committed:
Add test case for FormatScript, fix a bug with recent opcodes not being formated.

Summary: As per title.

Test Plan:
  make check

Added test case.

Reviewers: #bitcoin_abc, jasonbcox

Reviewed By: #bitcoin_abc, jasonbcox

Subscribers: teamcity

Differential Revision: https://reviews.bitcoinabc.org/D1620

Committer's notice: Stripped down to the essential changes for BU.

Co-authored-by: Awemany <awemany@protonmail.com>

deadalnix committed:
Activation test code for OP_CHECKDATASIG

[was: Add activation code for OP_CHECKDATASIG
 commit: 13eb8667a8073ee39f61039bbf3c7a172784a523]

Committer's notice: This is ABC's test code, not yet
working. Set commit author for attribution.

deadalnix committed:
Hash the input of OP_CHECKDATASIG instead of expecting the user to do it.

Summary:
This is safer

Co-authored-by: Amaury Séchet <deadalnix@gmail.com>
Co-authored-by: Jason B. Cox <contact@jasonbcox.com>

Test Plan: Updated tests to reflect the change.

Reviewers: #bitcoin_abc, Mengerian, schancel, jasonbcox

Reviewed By: #bitcoin_abc, Mengerian, schancel, jasonbcox

Subscribers: schancel, Mengerian, teamcity

Differential Revision: https://reviews.bitcoinabc.org/D1646

Committer's notice: Reworked for BU.

Co-authored-by: Awemany <awemany@protonmail.com>

deadalnix committed:
Use a single round of SHA256 on the CHECKDATASIG message

Summary: As per suggestion in the spec review.

Test Plan: Updated test cases.

Reviewers: #bitcoin_abc, jasonbcox, Mengerian

Reviewed By: #bitcoin_abc, jasonbcox, Mengerian

Subscribers: Mengerian, awemany, jasonbcox, teamcity

Differential Revision: https://reviews.bitcoinabc.org/D1653

Committer's notice: There was actually an earlier proposed
equivalent change to ABC, at https://reviews.bitcoinabc.org/D1652 .
This is the changeset from D1653 plus renaming 'message' to
'messagehash' in script/interpreter.cpp.

Co-authored-by: Awemany <awemany@protonmail.com>

awemany committed:
Check GetOpName(..) for completeness

awemany committed:
Remove OP_DATASIGVERIFY as it will be replaced

With the CHECKDATASIG and CHECKDATASIGVERIFY opcodes based on
DATASIGVERIFY.

awemany committed:
Implementation for InsecureRand256()

As this is a function used by ABC's variant of sigopcount_tests.cpp,
which will be in an upcoming commit.

jasonbcox committed:
Port of Bitcoin ABC's sighashtype.h file

Committer's notice:

This is a copy of Bitcoin ABC's SigHashType as of revision
f8283a3f284fc4722c1d6583b8746a17831d3bd0, meant to just aid in
porting test code. Notably, this should be the fixed variant wrt.
this issue:

https://github.com/mit-dci/cash-disclosure/blob/master/bitcoin-cash-disclosure-04252018.txt

(FIXME: Updated test code still missing!)

The test in test/script_sighashtype_tests.cpp has been ported as well.
The author in the commit line has been set for attribution, as well
as this Co-authored-by line:

Co-authored-by: Amaury Séchet <deadalnix@gmail.com>

awemany committed:
Import signature hash type declarations from interpreter.h

As sighashtype.h is meant only for testing for now.

deadalnix committed:
Import Bitcoin ABC's reworked signature encoding check function

Import this as IsValidSignatureEncodingWithoutSigHash(..). To be
used for CDS/-V.

This is the function that mainly resulted from this rework commit
in Bitcoin ABC:

49c7219314243398b9630ded78e97ebb0e3da776

Committer's notice: The author has been set to keep the attribution
of the author doing the rework. Also the signature has been changed
to work on vector<uint8_t>s again.

Co-authored-by: Awemany <awemany@protonmail.com>

awemany committed:
Replace exhaustive flags and hash type checking with randomization

This replaces the exhaustive enumeration of 1<<17 flags in
sigencoding_tests.cpp with a set of 100 random flag sets each,
as well as replacing the 2^8 different hash type checks with
16 random ones. This is to speed up the tests.

awemany committed:
Extend test_framework/script.py for CDS/-V

awemany committed:
Fix checkdatasig-activation.py for BU

Changes to get checkdatasig-activation.py work on the BU
test framework:

- The startup code seems to have assumed two nodes. Is just one
  node now (like it seems to be for ABC)

- There is no calculate_fee(..) RPC available in BU, this has
  been replaced with a simple ad-hoc fee calculation.

- Transaction hexlifying is done differently on BU

- Logging framework fixes

- BU rejects a CBlock in ConnectBlock with
  "bad-blk-signatures" wheras ABC rejects with "blk-bad-inputs"

- Add the checkdatasig RPC script to the test list

- Replace mentions of MAGNETIC_ANOMALY with NOV152018 (or similar)
  (as per comment from @sickpig)

sickpig commented on qa/rpc-tests/checkdatasig-activation.py:
```diff
@@ -0,0 +1,188 @@
+from test_framework.script import *
+
+# far into the future
+MAGNETIC_ANOMALY_START_TIME = 2000000000
```
I's rather use another name for this variable. `Magnetic Anomaly`  is the name ABC chose to name the November upgrade, it is peculiar to ABC way of doing things this is not something specified in the spec.

awemany replied:
Good point, will fix. `NOV2018FORK_START_TIME`?

sickpig replied:
@ptschip is using `NOV152018_START_TIME` in #1376

awemany replied:
Done.

sickpig commented on src/validation/validation.cpp:
```diff
@@ -1743,6 +1743,11 @@ uint32_t GetBlockScriptFlags(const CBlockIndex *pindex, const Consensus::Params
+    // BU: Set enableCheckDataSigVerify flag depending on whether CDS/-V should be enabled for this block
+    // This is a nasty side-effect kind of way of doing this and should be replaced
+    // with passing the flag to or setting it in a script verification class.
+    enableCheckDataSigVerify = IsNov152018Enabled(consensusparams, pindex->pprev);
```
Why `pindex->pprev`? `IsNov152018Enabled` already check if median time past (`MTP`) is grater or equal of activation time for current tip and tip - 1 (i.e. `pprev`). 
That way it checks for tip-2 and tip-1.

awemany replied:
That sounds right, thank you very much! I am not really firm in the matter of hard fork switch-over points and also I am confused by ABC's seemingly different logic.

As a further TODO, the boundary should be checked in a sharp sense and I had the `checkdatasig-activation.py` code passing, which it shouldn't given the right conditions, correct?

sickpig replied:
IMHO ABC and BU behave the same even if the way IsNov152018enabled and IsMagnetictAnomalyEnabled seem different.

As I said the former check from MTP trigger on both pindex and pindex-1 whereas the latter check only pindex this is why ABC is (probably, didn't check the code) passing pprev rather then the current tip.

awemany replied:
I changed it to use `chainActive.Tip()` now. I still need to reunderstand all this logic in more detail.

ptschip commented on src/script/sighashtype.h:
```diff
@@ -6,19 +6,11 @@
-    SIGHASH_ANYONECANPAY = 0x80,
-};
-
 /**
```
was there a problem in keeping them in sighashtype.h?  I think that it would make more sense to have removed them from  interpreter.h instead.

awemany replied:
To be honest, the chain split bug in ABC spooked me a bit and kept me from reworking everything to use `SigHashType` just yet. As I explained in my comments above, I did similarly for the signature encoding checks.

So far, `SigHashType` is *only* used for the test code that has been ported from ABC.

But I have no strong opinion on this and can of course rework this (now or later).

ptschip commented on qa/rpc-tests/checkdatasig-activation.py:
```diff
@@ -0,0 +1,186 @@
+        self.extra_args = [['-whitelist=127.0.0.1',
+                            "-magneticanomalyactivationtime=%d" % MAGNETIC_ANOMALY_START_TIME,
+                            "-replayprotectionactivationtime=%d" % (2 * MAGNETIC_ANOMALY_START_TIME)]]
+
```
you can get rid of the -replayprotectionactivationtime and change the -magenticanomalyactivation time to BU's version of the activation time.

awemany replied:
Hmm, this must be code from the midst of my changeset? Because it already is adapted accordingly at the HEAD commit of this PR branch.

ptschip commented on src/test/checkdatasig_tests.cpp:
```diff
@@ -0,0 +1,255 @@
+                       const CScript &script, ScriptError expected) {
+    enableCheckDataSigVerify = (flags & SCRIPT_ENABLE_CHECKDATASIG);
+    BOOST_TEST_INFO("CheckError()");
+    BOOST_TEST_INFO(tfm::format("Flags: %x, enCDS: %d, origStack: %s, Script: %s, Expected Error: %s",
```
BOOST_TEST_INFO is not compatible with our current min boost level that we support.

awemany replied:
Good catch. I guess I can just remove this code for now, it is from debugging.

ptschip commented:
fair enough , can do it some other time then

On 08/10/2018 2:09 PM, awemany wrote:
>
> *@awemany* commented on this pull request.
>
> ------------------------------------------------------------------------
>
> In src/script/sighashtype.h
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/1390#discussion_r223497978>:
>
> >  
>  #include <cstdint>
>  #include <stdexcept>
>  
> -/** Signature hash types/flags */
> -enum {
> -    SIGHASH_ALL = 1,
> -    SIGHASH_NONE = 2,
> -    SIGHASH_SINGLE = 3,
> -    SIGHASH_FORKID = 0x40,
> -    SIGHASH_ANYONECANPAY = 0x80,
> -};
> -
>  /**
>
> To be honest, the chain split bug in ABC spooked me a bit and kept me
> from reworking everything to use |SigHashType| just yet. As I
> explained in my comments above, I did similarly for the signature
> encoding checks.
>
> So far, |SigHashType| is /only/ used for the test code that has been
> ported from ABC.
>
> But I have no strong opinion on this and can of course rework this
> (now or later).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/1390#discussion_r223497978>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AMRF0G-FK1TLbwcrIc0QULA8v6OPgqV-ks5ui77vgaJpZM4XL1vZ>.
>

sickpig commented on src/script/interpreter.h:
```diff
@@ -92,10 +92,15 @@ enum
 
+    // Public keys in scripts must be compressed
+    //
+    SCRIPT_VERIFY_COMPRESSED_PUBKEYTYPE = (1U << 15),
```
nit: move `SCRIPT_VERIFY_COMPRESSED_PUBKEYTYPE` before `SCRIPT_ENABLE_SIGHASH_FORKID`

awemany replied:
Done.

sickpig reviewed (commented):
In e1c5a10 commit message you repeated `Co-authored-by`

sickpig commented on src/script/interpreter.h:
```diff
@@ -207,8 +213,10 @@ bool VerifyScript(const CScript &scriptSig,
     unsigned char *sighashtype = NULL);
 
-// string prefixed to data when validating signed messages either via DATASIGVERIFY or RPC call.  This ensures
+// string prefixed to data when validating signed messages  via RPC call.  This ensures
```
double space between `messages` and `via`.

awemany replied:
Done.

sickpig commented on src/test/sigencoding_tests.cpp:
```diff
@@ -0,0 +1,389 @@
+// Copyright (c) 2018 The Bitcoin developers
```
need to add `sigencoding_tests.cpp` to `.formatted-files`

awemany replied:
Done. Also added the other missing files. See bottom "Cosmetics" commit.

awemany committed:
Activation code for CHECKDATASIG/-VERIFY

awemany committed:
Cosmetics

- reorg script flags to be in order, fix space
  in interpreter.h (as per @sickpig review nits)

- add sigencoding_tests.cpp to .formatted-files
  (as per @sickpig)
- also add checkdatasig_tests.cpp, sighashtype_tests.cpp
  to .formatted-files
- Reformat the above, inserting 'clang-format' markers wher
  appropriate

awemany commented:
Ok, reworked.

A couple points:

- It is using the script flag activation mechanism now and no more global variable.

- *Please* double and triple check the activation code. I am unsure whether I put the code in the right places and also whether the block height logic is correct. I still need to double check myself. Furthermore, if I am not mistaken, with the current logic, BU will ignore CDS transactions into the mempool until the fork has already taken place. Which I guess is o.k.?

- This is not to mean to not check the other code ...

- The other code where I deviated considerable from ABC is the way signature checking for CDS/-V has been "bolted on". As I said above in the convo with @ptschip, I tried to keep the existing code intact as much as as possible. But the resulting smaller changeset is still consensus-critical ...

- Squashing: I can squash however you guys want but more than half of the commits are imported. So if you want something squashed, please tell me ASAP.

gandrewstone commented on src/script/interpreter.cpp:
```diff
@@ -224,37 +403,68 @@ static bool IsDefinedHashtypeSignature(const valtype &vchSig)
+static bool CheckSignatureEncodingSigHashChoice(const vector<unsigned char> &vchSig,
+    unsigned int flags,
+    ScriptError *serror,
+    const bool check_sighash)
```
couldn't check_sighash be part of flags?

EDIT: I suppose it doesn't matter much since this is not part of the external API

gandrewstone commented on src/script/interpreter.cpp:
```diff
@@ -1259,6 +1431,63 @@ bool EvalScript(vector<vector<unsigned char> > &stack,
+                    {
+                        if (fSuccess)
+                        {
+                            popstack(stack);
```
I seems unnecessary to push true or false onto the stack only to pop it back off if VERIFY.  Why not use:

if (opcode == CDS) stack.push_back()
else if (!fSuccess) // opcode == CDSV
  {
    return set_error(...)
  }

gandrewstone commented on src/script/sighashtype.h:
```diff
@@ -0,0 +1,82 @@
+// Copyright (c) 2017-2018 Bitcoin developers
```
Add file to .formatted-files

gandrewstone commented on src/script/sign.h:
```diff
@@ -103,7 +103,7 @@ std::vector<unsigned char> signmessage(const BYTEARRAY &data, const CKey &key)
 }
 
-/** sign arbitrary data using the same algorithm as the signmessage/verifymessage RPCs and OP_DATASIGVERIFY */
+/** sign arbitrary data using the same algorithm as the signmessage/verifymessage RPCs and OP_CHECKDATASIG(VERIFY) */
```
its CDS doesn't use the pubkey recoverable signature does it?  so AFAIK we need to delete "...and OP_CHECKDATASIG(VERIFY)"

gandrewstone commented:
To optimize rewind, we now just place uncommitted tx into the processing queues rather than ensuring that these tx fully enter the mempool.  This allows the multithreaded tx processing pipeline to deal with them rather than the rewind logic.  However, this means that you may not be guaranteed that tx have re-entered the mempool when invalidateblock completes.  This is not an issue for real use, but messes up some tests. We may want to add a "flush" RPC call but for now to make sure the test consistently succeeds, you need to sleep after invalidateblock.  3 seconds is incredibly generous (see diff below), but travis can be incredibly slow.

diff --git a/qa/rpc-tests/checkdatasig-activation.py b/qa/rpc-tests/checkdatasig-activation.py
index 2a1ecd9..d1bc6a4 100755
--- a/qa/rpc-tests/checkdatasig-activation.py
+++ b/qa/rpc-tests/checkdatasig-activation.py
@@ -178,9 +178,11 @@ class CheckDataSigActivationTest(ComparisonTestFramework):
         assert(tx0id not in set(node.getrawmempool()))
 
         node.invalidateblock(format(nov152018forkblock.sha256, 'x'))
+        time.sleep(3) # wait for tx move back through the processing queues
         assert(tx0id in set(node.getrawmempool()))
 
         node.invalidateblock(format(fork_block.sha256, 'x'))
+        time.sleep(3) # wait for tx move back through the processing queues
         assert(tx0id not in set(node.getrawmempool()))

sickpig committed:
Wait for 3 seconds after invalidateblock() before checking for assertions

Add 3 sec waitFor() after invalidateblock(), this is mainly done to avoid Travis
failures. We are going to probably remove it once we move to gitlab.com

gandrewstone commented:
timing out travis (as is common) but working fine over multiple runs on my computer
